### PR TITLE
Cross tile index performance

### PR DESCRIPTION
**Pull request by pozdnyakov**
Before this change cross tile symbol indexing was a performance bottle neck, for the overscaled tiles in particular. This pull request optimizes the cross tile symbol indexing code path:
- find matches only among the buckets that have the same leader Id.
- for overscaled tiles, index only the symbols inside the viewport
- multiple minor optimizations and code simplifications

Fixes https://github.com/mapbox/mapbox-gl-native-team/issues/133

### CODE REVIEW

File: src/mbgl/text/cross_tile_symbol_index.cpp [hunk 1]
```diff
@@ -112,8 +131,28 @@ bool CrossTileSymbolLayerIndex::addBucket(const OverscaledTileID& tileID,
...
 
-    for (auto& symbolInstance: bucket.symbolInstances) {
-        symbolInstance.crossTileID = 0;
+    bucket.hasUninitializedSymbols = false;
+
+    if (tileID.overscaleFactor() > 1u) {
+        // For overscaled tiles the viewport might be showing only a small part of the tile,
+        // so we filter out the off-screen symbols to improve the performance.
+        mat4 posMatrix;
+        const auto& projMatrix = transformState.getProjectionMatrix();
```
Comment (astojilj) [inline src/mbgl/text/cross_tile_symbol_index.cpp]: Reconsider reusing [renderTile.matrix calculation](https://github.com/mapbox/mapbox-gl-native/blob/c7bbd5bbef0780df5d0a5498e14051ee63c24e44/src/mbgl/renderer/render_tile.cpp#L124) and passing it here from `CrossTileSymbolIndex::addLayer`, also in https://github.com/mapbox/mapbox-gl-native/blob/9bb1f898030a8bf73c11e53f498684e24a61ce60/src/mbgl/text/placement.cpp#L157 as all three might calculate the same.

Comment (pozdnyakov) [reply src/mbgl/text/cross_tile_symbol_index.cpp]: Great point, thanks!

File: src/mbgl/text/cross_tile_symbol_index.cpp [hunk 2]
```diff
@@ -95,13 +95,32 @@ void CrossTileSymbolLayerIndex::handleWrapJump(float newLng) {
     lng = newLng;
 }
 
+namespace {
+
+bool isInVewport(const mat4& posMatrix, const Point<float>& point) {
+    vec4 p = {{point.x, point.y, 0, 1}};
+    matrix::transformMat4(p, p, posMatrix);
+
+    constexpr float buffer = 0.5f;
```
Comment (astojilj) [inline src/mbgl/text/cross_tile_symbol_index.cpp]: A nit: reconsider:
```
// buffer covers area of the next zoom level (current zoom - 1 covered area).
constexpr float buffer = 1.0f;
float x = p[0] / p[3];
float y = p[1] / p[3];
return (x > -1 - buffer && y > -1 - buffer && x < 1 + buffer && y < 1 + buffer);
```

File: src/mbgl/text/cross_tile_symbol_index.cpp [hunk 3]
```diff
@@ -101,24 +129,41 @@ bool CrossTileSymbolLayerIndex::addBucket(const OverscaledTileID& tileID, Symbol
...
+    bucket.hasUninitializedSymbols = false;
+
+    if (tileID.overscaleFactor() > 1u) {
+        // For overscaled tiles the viewport might be showing only a small part of the tile,
+        // so we filter out the off-screen symbols to improve the performance.
+        for (auto& symbolInstance : bucket.symbolInstances) {
+            if (isInVewport(tileMatrix, symbolInstance.anchor.point)) {
+                symbolInstance.crossTileID = 0u;
+            } else {
+                symbolInstance.crossTileID = SymbolInstance::invalidCrossTileID();
```
Comment (alexshalamov) [inline src/mbgl/text/cross_tile_symbol_index.cpp]: Would this delay symbol placement by 300ms for out-of-viewport symbols when we pan overscaled tile?

Comment (pozdnyakov) [reply src/mbgl/text/cross_tile_symbol_index.cpp]: it won't, as we reassign those ids at every frame

File: src/mbgl/text/cross_tile_symbol_index.cpp [hunk 4]
```diff
@@ -86,11 +96,29 @@ void CrossTileSymbolLayerIndex::handleWrapJump(float newLng) {
     lng = newLng;
 }
 
-bool CrossTileSymbolLayerIndex::addBucket(const OverscaledTileID& tileID, SymbolBucket& bucket, uint32_t& maxCrossTileID) {
-    const auto& thisZoomIndexes = indexes[tileID.overscaledZ];
+namespace {
+
+bool isInVewport(const mat4& posMatrix, const Point<float>& point) {
```
Comment (springmeyer) [inline src/mbgl/text/cross_tile_symbol_index.cpp]: @pozdnyakov should this be spelled `isInViewport`?

Comment (pozdnyakov) [reply src/mbgl/text/cross_tile_symbol_index.cpp]: yeah :grimacing: thanks for the catch, will fix it in a following patch..

**Pull request merged**